src/shims/fs.rs

```diff
@@ -160,20 +171,26 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             return Ok(0);
         }
         let fd = this.read_scalar(fd_op)?.to_i32()?;
-        let buf_scalar = this.read_scalar(buf_op)?.not_undef()?;
-
-        // Remove the file handle to avoid borrowing issues.
-        this.remove_handle_and(fd, |mut handle, this| {
-            // Don't use `?` to avoid returning before reinserting the handle.
-            let bytes = this.force_ptr(buf_scalar).and_then(|buf| {
-                this.memory
-                    .get_mut(buf.alloc_id)?
-                    .get_bytes_mut(&*this.tcx, buf, Size::from_bytes(count))
-                    .map(|buffer| handle.file.read(buffer))
-            });
-            this.machine.file_handler.handles.insert(fd, handle).unwrap_none();
-            this.try_unwrap_io_result(bytes?.map(|bytes| bytes as i64))
-        })
+        let buf = this.read_scalar(buf_op)?.not_undef()?;
+
+        if let Some(handle) = this.machine.file_handler.handles.get_mut(&fd) {
+            // We want to read at most `count` bytes
+            let mut bytes = vec![0; count as usize];
+            let result = handle.file.read(&mut bytes);
+
+            if let Ok(c) = result {
+                // Check that we read less than `i64::MAX` bytes.
+                if c > (i64::max_value() as usize) {
```

Comment:
That cast will not work out... rather cast them both to `u64` for comparison.

Reply:
But I think this check is really not needed (as you said before I think); we have the call `count` many bytes which is already sufficiently bounded.

The weird part is that `count` is a `usize` but later you make it an `i64`... why that? What are the relevant types on the interpreted machine?

Reply:
the signature for [`read`](http://man7.org/linux/man-pages/man2/read.2.html) is:
```c
ssize_t read(int fd, void *buf, size_t count);
```
So, the `count` argument is an `usize` but the return type is `isize`. I have no idea what should happen if someone passes to count a value larger than `isize::max_value()`

Reply:
These are *target* `usize` though. So e.g. on a 32bit host with a 64bit target, these cannot be loslessly converted to *host* `usize`.

Target `usize` is `u64` in Miri. Ideally we'd use `TryInto` to convert that into `usize`, but unfortunately those instances do not exist. So instead I suggest adding helper methods like
```rust
fn try_into_usize(i: impl Into<u128>) -> Option<usize>
{
  let i: u128 = i;
  if i > usize::max_value() as u128 { None } else { Some(i as usize) }
}
fn try_from_usize<T: TryFrom<u128>>(i: usize) -> Option<T>
{
  T::try_from(u128::from(i))
}
```

Reply:
Hm, that will be really confusing with the scalar `to_usize` though.^^

Maybe `try_into/from_host_usize`?

Reply:
OK I'll add them :)

Reply:
I added the functions and I'm using them in `read` and `write` (see https://github.com/rust-lang/miri/pull/1022/commits/56c5e53553859d153009f76d2da15bf2d0e3d037). I don't like how it looks but it works. Maybe we can reformulate `try_unwrap_io_result` somehow.